pstoreds/peerstore.go

```diff
@@ -42,6 +43,59 @@ func NewPeerstore(ctx context.Context, store ds.TxnDatastore, opts Options) (pst
 		return nil, err
 	}
 
-	ps := pstore.NewPeerstore(pstoremem.NewKeyBook(), addrBook, pstoremem.NewPeerMetadata())
+	keyBook, err := NewKeyBook(ctx, store, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	peerMetadata, err := NewPeerMetadata(ctx, store, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	ps := pstore.NewPeerstore(keyBook, addrBook, peerMetadata)
 	return ps, nil
 }
+
+// uniquePeerIds extracts and returns unique peer IDs from database keys.
+func uniquePeerIds(ds ds.TxnDatastore, prefix ds.Key, extractor func(result query.Result) string) (peer.IDSlice, error) {
+	var (
+		q       = query.Query{Prefix: prefix.String(), KeysOnly: true}
+		results query.Results
+		err     error
+	)
+
+	txn, err := ds.NewTransaction(true)
+	if err != nil {
+		return peer.IDSlice{}, err
+	}
+	defer txn.Discard()
+
+	if results, err = txn.Query(q); err != nil {
+		log.Error(err)
+		return peer.IDSlice{}, err
+	}
+
+	defer results.Close()
+
+	idset := make(map[string]struct{})
+	for result := range results.Next() {
+		k := extractor(result)
+		idset[k] = struct{}{}
+		//key := ds.RawKey(result.Key)
+		//idset[key.Parent().Name()] = struct{}{}
+	}
+
+	if len(idset) == 0 {
+		return peer.IDSlice{}, nil
+	}
+
+	ids := make(peer.IDSlice, len(idset))
```

Comment:
we can do this all in one loop actually, checking set membership at top of loop, continuing if found, otherwise adding to our return results and inserting into idset

Reply:
Yeah, but memory-wise it's nicer to allocate a slice of a known size, than to start from 0 and keep growing.

Reply:
why not just allocate the full size w/ initial length 0?

Reply:
i suppose the computation is basically free in this IO bound thing so this is fine

Reply:
We can't allocate a capacity because we don't know the count of unique IDs, nor the count of entries in the iterator. So we'd start from 0, and the worst case scenario is really bad. Iterating twice is a fair tradeoff I believe.